messaging/src/main/java/org/axonframework/eventhandling/processors/streaming/token/store/inmemory/InMemoryTokenStore.java

```diff
@@ -50,90 +56,125 @@ public class InMemoryTokenStore implements TokenStore {
+                        If the use of an in memory token store is intentional, this warning can be ignored.
+                        If the tokens should be persisted, use the JPA, JDBC or MongoDB token store instead.
+                        """
+        );
     }
 
+    @Nonnull
     @Override
-    public void initializeTokenSegments(@Nonnull String processorName, int segmentCount)
-            throws UnableToClaimTokenException {
-        initializeTokenSegments(processorName, segmentCount, null);
-    }
-
-    @Override
-    public void initializeTokenSegments(@Nonnull String processorName, int segmentCount, TrackingToken initialToken)
-            throws UnableToClaimTokenException {
-        if (fetchSegments(processorName).length > 0) {
-            throw new UnableToClaimTokenException("Could not initialize segments. Some segments were already present.");
-        }
-        for (int segment = 0; segment < segmentCount; segment++) {
-            tokens.put(new ProcessAndSegment(processorName, segment), getOrDefault(initialToken, NULL_TOKEN));
-        }
+    public CompletableFuture<Void> initializeTokenSegments(
+            @Nonnull String processorName,
+            int segmentCount,
+            @Nullable TrackingToken initialToken,
+            @Nullable ProcessingContext context
+    ) throws UnableToClaimTokenException {
+        return fetchSegments(processorName, context)
+                .thenAccept(segments -> {
+                    if (segments.length > 0) {
+                        throw new UnableToClaimTokenException(
+                                "Could not initialize segments. Some segments were already present.");
+                    }
+                    for (int segment = 0; segment < segmentCount; segment++) {
+                        tokens.put(new ProcessAndSegment(processorName, segment),
+                                   getOrDefault(initialToken, NULL_TOKEN));
+                    }
+                });
     }
 
-    // TODO #3432 - CurrentUnitOfWork is here to mimic transactional behavior, adjust it by mean of ProcessingContext
+    @Nonnull
     @Override
-    public void storeToken(TrackingToken token, @Nonnull String processorName, int segment) {
-        if (CurrentUnitOfWork.isStarted()) {
-            CurrentUnitOfWork.get().afterCommit(uow -> tokens.put(new ProcessAndSegment(processorName, segment),
-                                                                  getOrDefault(token, NULL_TOKEN)));
+    public CompletableFuture<Void> storeToken(@Nullable TrackingToken token,
+                                              @Nonnull String processorName,
+                                              int segment,
+                                              @Nullable ProcessingContext context) {
+        Objects.requireNonNull(context, "processingContext may not be null for an InMemoryTokenStore");
+        if (context.isStarted()) {
+            var future = runAsync(() -> tokens.put(
```

Comment:
The `ProcessingContext` will under the hood ensure things are ran asynchronous. So, you can simply pass the operation inside the `context#onAfterCommit()` hook!

Reply:
yes it's better, because then in the UnitOfWork we can define the ExecutorService, here it's even not defined on which Thread it will be run.